File: backend/internal/observability/observability.go
package observability

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	dbobs "github.com/kurodakayn/mpp-backend/internal/db"
	"github.com/labstack/echo/v4"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/collectors"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

const (
	requestIDHeader                   = "X-Request-ID"
	traceIDHeader                     = "X-Trace-ID"
	databaseSlowQueryThresholdEnv     = "DB_SLOW_QUERY_THRESHOLD"
	defaultDatabaseSlowQueryThreshold = 250 * time.Millisecond
)

type contextKey string

const traceIDContextKey contextKey = "trace_id"

type Suite struct {
	serviceName      string
	registry         *prometheus.Registry
	requests         *prometheus.CounterVec
	duration         *prometheus.HistogramVec
	inFlight         *prometheus.GaugeVec
	info             *prometheus.GaugeVec
	databaseObserver *DatabaseQueryObserver
}

type DatabaseQueryObserver struct {
	serviceName   string
	slowThreshold time.Duration
	queries       *prometheus.CounterVec
	duration      *prometheus.HistogramVec
	slowQueries   *prometheus.CounterVec
}

type requestLog struct {
	Time      string  `json:"time"`
	Service   string  `json:"service"`
	TraceID   string  `json:"trace_id"`
	RequestID string  `json:"request_id"`
	Method    string  `json:"method"`
	Path      string  `json:"path"`
	Route     string  `json:"route"`
	Status    int     `json:"status"`
	LatencyMS float64 `json:"latency_ms"`
	RemoteIP  string  `json:"remote_ip"`
	UserAgent string  `json:"user_agent"`
	BytesIn   int64   `json:"bytes_in"`
	BytesOut  int64   `json:"bytes_out"`
	Error     string  `json:"error,omitempty"`
}

type databaseSlowQueryLog struct {
	Time         string  `json:"time"`
	Service      string  `json:"service"`
	TraceID      string  `json:"trace_id,omitempty"`
	Operation    string  `json:"operation"`
	Table        string  `json:"table"`
	Status       string  `json:"status"`
	QueryHash    string  `json:"query_hash"`
	DurationMS   float64 `json:"duration_ms"`
	RowsAffected int64   `json:"rows_affected"`
	SQL          string  `json:"sql"`
	Error        string  `json:"error,omitempty"`
}

func New(serviceName string) *Suite {
	serviceName = strings.TrimSpace(serviceName)
	if serviceName == "" {
		serviceName = "mpp-service"
	}

	requests := prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "mpp_http_requests_total",
		Help: "Total HTTP requests served by MPP services.",
	}, []string{"service", "method", "route", "status"})
	duration := prometheus.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "mpp_http_request_duration_seconds",
		Help:    "HTTP request duration by service, method, route, and status.",
		Buckets: []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10},
	}, []string{"service", "method", "route", "status"})
	inFlight := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "mpp_http_in_flight_requests",
		Help: "Current in-flight HTTP requests by MPP service.",
	}, []string{"service"})
	info := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "mpp_service_info",
		Help: "Static information marker for an MPP service.",
	}, []string{"service"})
	dbQueries := prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "mpp_db_queries_total",
		Help: "Total database queries executed by MPP services.",
	}, []string{"service", "operation", "table", "status"})
	dbDuration := prometheus.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "mpp_db_query_duration_seconds",
		Help:    "Database query duration by service, operation, table, and status.",
		Buckets: []float64{0.001, 0.0025, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5},
	}, []string{"service", "operation", "table", "status"})
	dbSlowQueries := prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "mpp_db_slow_queries_total",
		Help: "Total database queries that exceeded the configured slow query threshold.",
	}, []string{"service", "operation", "table", "status"})

	registry := prometheus.NewRegistry()
	registry.MustRegister(
		collectors.NewGoCollector(),
		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
		requests,
		duration,
		inFlight,
		info,
		dbQueries,
		dbDuration,
		dbSlowQueries,
	)
	info.WithLabelValues(serviceName).Set(1)
	databaseObserver := &DatabaseQueryObserver{
		serviceName:   serviceName,
		slowThreshold: databaseSlowQueryThresholdFromEnv(),
		queries:       dbQueries,
		duration:      dbDuration,
		slowQueries:   dbSlowQueries,
	}

	return &Suite{
		serviceName:      serviceName,
		registry:         registry,
		requests:         requests,
		duration:         duration,
		inFlight:         inFlight,
		info:             info,
		databaseObserver: databaseObserver,
	}
}

func (s *Suite) RegisterRoutes(e *echo.Echo) {
	e.GET("/metrics", echo.WrapHandler(promhttp.HandlerFor(s.registry, promhttp.HandlerOpts{})))
}

func (s *Suite) DatabaseQueryObserver() *DatabaseQueryObserver {
	return s.databaseObserver
}

func (s *Suite) Middleware() echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			req := c.Request()
			traceID := traceIDFromRequest(req)
			req = req.WithContext(ContextWithTraceID(req.Context(), traceID))
			c.SetRequest(req)
			c.Set("trace_id", traceID)
			c.Set("request_id", traceID)
			c.Response().Header().Set(requestIDHeader, traceID)
			c.Response().Header().Set(traceIDHeader, traceID)

			startedAt := time.Now()
			s.inFlight.WithLabelValues(s.serviceName).Inc()
			err := next(c)
			if err != nil {
				c.Error(err)
			}
			s.inFlight.WithLabelValues(s.serviceName).Dec()

			status := c.Response().Status
			if status == 0 {
				status = http.StatusOK
			}
			route := routePath(c)
			statusLabel := strconv.Itoa(status)
			latency := time.Since(startedAt)
			s.requests.WithLabelValues(s.serviceName, req.Method, route, statusLabel).Inc()
			s.duration.WithLabelValues(s.serviceName, req.Method, route, statusLabel).Observe(latency.Seconds())

			s.logRequest(c, traceID, route, status, latency, err)
			return nil
		}
	}
}

func ContextWithTraceID(ctx context.Context, traceID string) context.Context {
	if ctx == nil {
		ctx = context.Background()
	}
	traceID = strings.TrimSpace(traceID)
	if traceID == "" {
		return ctx
	}
	return context.WithValue(ctx, traceIDContextKey, traceID)
}

func TraceIDFromContext(ctx context.Context) string {
	if ctx == nil {
		return ""
	}
	traceID, _ := ctx.Value(traceIDContextKey).(string)
	return strings.TrimSpace(traceID)
}

func traceIDFromRequest(req *http.Request) string {
	if value := strings.TrimSpace(req.Header.Get(requestIDHeader)); value != "" {
		return value
	}
	if value := strings.TrimSpace(req.Header.Get(traceIDHeader)); value != "" {
		return value
	}
	return uuid.NewString()
}

func routePath(c echo.Context) string {
	if path := strings.TrimSpace(c.Path()); path != "" {
		return path
	}
	if c.Request() != nil && c.Request().URL != nil {
		return c.Request().URL.Path
	}
	return "unknown"
}

func (s *Suite) logRequest(c echo.Context, traceID, route string, status int, latency time.Duration, err error) {
	req := c.Request()
	event := requestLog{
		Time:      time.Now().UTC().Format(time.RFC3339Nano),
		Service:   s.serviceName,
		TraceID:   traceID,
		RequestID: traceID,
		Method:    req.Method,
		Path:      req.URL.Path,
		Route:     route,
		Status:    status,
		LatencyMS: float64(latency.Microseconds()) / 1000,
		RemoteIP:  c.RealIP(),
		UserAgent: req.UserAgent(),
		BytesIn:   req.ContentLength,
		BytesOut:  c.Response().Size,
	}
	if err != nil {
		event.Error = err.Error()
	}
	payload, marshalErr := json.Marshal(event)
	if marshalErr != nil {
		log.Printf("observability request log marshal failed: %v", marshalErr)
		return
	}
	if _, err := os.Stdout.Write(append(payload, '\n')); err != nil {
		log.Printf("observability request log write failed: %v", err)
	}
}

func (o *DatabaseQueryObserver) ObserveQuery(ctx context.Context, observation dbobs.QueryObservation) {
	if o == nil {
		return
	}

	operation := metricLabel(observation.Operation, "unknown")
	table := metricLabel(observation.Table, "unknown")
	status := "ok"
	if observation.Err != nil {
		status = "error"
	}

	durationSeconds := observation.Duration.Seconds()
	if durationSeconds < 0 {
		durationSeconds = 0
	}

	o.queries.WithLabelValues(o.serviceName, operation, table, status).Inc()
	o.duration.WithLabelValues(o.serviceName, operation, table, status).Observe(durationSeconds)

	if o.slowThreshold <= 0 || observation.Duration < o.slowThreshold {
		return
	}

	o.slowQueries.WithLabelValues(o.serviceName, operation, table, status).Inc()
	o.logSlowQuery(ctx, observation, operation, table, status)
}

func (o *DatabaseQueryObserver) logSlowQuery(ctx context.Context, observation dbobs.QueryObservation, operation, table, status string) {
	event := databaseSlowQueryLog{
		Time:         time.Now().UTC().Format(time.RFC3339Nano),
		Service:      o.serviceName,
		TraceID:      TraceIDFromContext(ctx),
		Operation:    operation,
		Table:        table,
		Status:       status,
		QueryHash:    observation.QueryHash,
		DurationMS:   float64(observation.Duration.Microseconds()) / 1000,
		RowsAffected: observation.RowsAffected,
		SQL:          observation.SQL,
	}
	if observation.Err != nil {
		event.Error = observation.Err.Error()
	}

	payload, marshalErr := json.Marshal(event)
	if marshalErr != nil {
		log.Printf("observability database slow query log marshal failed: %v", marshalErr)
		return
	}
	if _, err := os.Stdout.Write(append(payload, '\n')); err != nil {
		log.Printf("observability database slow query log write failed: %v", err)
	}
}

func metricLabel(value, fallback string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return fallback
	}
	if len(value) > 96 {
		return value[:96]
	}
	return value
}

func databaseSlowQueryThresholdFromEnv() time.Duration {
	raw := strings.TrimSpace(os.Getenv(databaseSlowQueryThresholdEnv))
	if raw == "" {
		return defaultDatabaseSlowQueryThreshold
	}
	threshold, err := time.ParseDuration(raw)
	if err != nil || threshold < 0 {
		return defaultDatabaseSlowQueryThreshold
	}
	return threshold
}

File: ai-service/observability.py
import json
import time
from uuid import uuid4

from fastapi import FastAPI, Request, Response
from prometheus_client import (
    CONTENT_TYPE_LATEST,
    Counter,
    Gauge,
    Histogram,
    generate_latest,
)

REQUEST_ID_HEADER = "x-request-id"
TRACE_ID_HEADER = "x-trace-id"
SERVICE_NAME = "ai-service"

REQUESTS = Counter(
    "mpp_http_requests_total",
    "Total HTTP requests served by MPP services.",
    ("service", "method", "route", "status"),
)
DURATION = Histogram(
    "mpp_http_request_duration_seconds",
    "HTTP request duration by service, method, route, and status.",
    ("service", "method", "route", "status"),
    buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10),
)
IN_FLIGHT = Gauge(
    "mpp_http_in_flight_requests",
    "Current in-flight HTTP requests by MPP service.",
    ("service",),
)
INFO = Gauge(
    "mpp_service_info",
    "Static information marker for an MPP service.",
    ("service",),
)
INFO.labels(SERVICE_NAME).set(1)


def configure_observability(app: FastAPI) -> None:
    @app.middleware("http")
    async def request_observability(request: Request, call_next):
        trace_id = _trace_id_from_request(request)
        request.state.trace_id = trace_id
        request.state.request_id = trace_id
        started_at = time.perf_counter()
        status_code = 500
        error = ""
        response = None

        IN_FLIGHT.labels(SERVICE_NAME).inc()
        try:
            response = await call_next(request)
            status_code = response.status_code
            response.headers[REQUEST_ID_HEADER] = trace_id
            response.headers[TRACE_ID_HEADER] = trace_id
            return response
        except Exception as exc:
            error = str(exc)
            raise
        finally:
            IN_FLIGHT.labels(SERVICE_NAME).dec()
            route = _route_path(request)
            latency = time.perf_counter() - started_at
            REQUESTS.labels(SERVICE_NAME, request.method, route, str(status_code)).inc()
            DURATION.labels(
                SERVICE_NAME, request.method, route, str(status_code)
            ).observe(latency)
            _log_request(
                request, trace_id, route, status_code, latency, response, error
            )

    @app.get("/metrics", include_in_schema=False)
    async def metrics() -> Response:
        return Response(generate_latest(), media_type=CONTENT_TYPE_LATEST)


def _trace_id_from_request(request: Request) -> str:
    request_id = request.headers.get(REQUEST_ID_HEADER, "").strip()
    if request_id:
        return request_id
    trace_id = request.headers.get(TRACE_ID_HEADER, "").strip()
    if trace_id:
        return trace_id
    return str(uuid4())


def _route_path(request: Request) -> str:
    route = request.scope.get("route")
    route_path = getattr(route, "path", None)
    return route_path or request.url.path or "unknown"


def _log_request(
    request: Request,
    trace_id: str,
    route: str,
    status_code: int,
    latency: float,
    response,
    error: str,
) -> None:
    event = {
        "time": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "service": SERVICE_NAME,
        "trace_id": trace_id,
        "request_id": trace_id,
        "method": request.method,
        "path": request.url.path,
        "route": route,
        "status": status_code,
        "latency_ms": round(latency * 1000, 3),
        "remote_ip": request.client.host if request.client else "",
        "user_agent": request.headers.get("user-agent", ""),
        "bytes_in": _content_length(request),
        "bytes_out": _response_size(response),
    }
    if error:
        event["error"] = error
    print(json.dumps(event, separators=(",", ":")), flush=True)


def _content_length(request: Request) -> int:
    value = request.headers.get("content-length", "").strip()
    if not value:
        return 0
    try:
        return int(value)
    except ValueError:
        return 0


def _response_size(response) -> int:
    if response is None:
        return 0
    value = response.headers.get("content-length", "").strip()
    if not value:
        return 0
    try:
        return int(value)
    except ValueError:
        return 0
